ksqldb-engine/src/main/java/io/confluent/ksql/internal/ThroughputMetricsReporter.java

```diff
@@ -0,0 +1,260 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.apache.kafka.common.utils.Utils.mkSet;
+
+public class ThroughputMetricsReporter implements MetricsReporter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ThroughputMetricsReporter.class);
+    private static final String RECORDS_CONSUMED = "records-consumed";
+    private static final String BYTES_CONSUMED = "bytes-consumed";
+    private static final String RECORDS_PRODUCED = "records-produced";
+    private static final String BYTES_PRODUCED = "bytes-produced";
+    private static final Set<String> THROUGHPUT_METRIC_NAMES =
+        mkSet(RECORDS_CONSUMED, BYTES_CONSUMED, RECORDS_PRODUCED, BYTES_PRODUCED);
+    private static final Pattern NAMED_TOPOLOGY_PATTERN = Pattern.compile("(.*?)__\\d*_\\d*");
+    private static final Pattern QUERY_ID_PATTERN =
+        Pattern.compile("(?<=query_|transient_)(.*?)(?=-)");
+
+    // CHECKSTYLE_RULES.OFF: LineLength
+    //
+    private static final Map<String, Map<String, Map<MetricName, ThroughputTotalMetric>>> registeredMetrics = new HashMap<>();
+    // CHECKSTYLE_RULES.ON: LineLength
+    private static final Map<String, String> customTags = new HashMap<>();
+    private Metrics metricRegistry;
+
+    @Override
+    public void init(final List<KafkaMetric> initial) {
+    }
+
+    @Override
+    public synchronized void configure(final Map<String, ?> configMap) {
+        this.metricRegistry = (Metrics) requireNonNull(
+            configMap.get(KsqlConfig.KSQL_INTERNAL_METRICS_CONFIG)
+        );
+        customTags.putAll(KsqlConfig.getStringAsMap(KsqlConfig.KSQL_CUSTOM_METRICS_TAGS, configMap));
+    }
+
+    @Override
+    public void metricChange(final KafkaMetric metric) {
+        if (!THROUGHPUT_METRIC_NAMES.contains(metric.metricName().name())) {
+            return;
+        }
+        addMetric(
+            metric,
+            getQueryId(metric),
+            getTopic(metric)
+        );
+    }
+
+    @Override
+    public void metricRemoval(final KafkaMetric metric) {
+        if (!THROUGHPUT_METRIC_NAMES.contains(metric.metricName().name())) {
+            return;
+        }
+
+        removeMetric(
+            metric,
+            getQueryId(metric),
+            getTopic(metric)
+        );
+    }
+
+    @Override
+    public void close() {
+    }
+
+    @Override
+    public Set<String> reconfigurableConfigs() {
+        return null;
+    }
+
+    @Override
+    public void validateReconfiguration(final Map<String, ?> configs) throws ConfigException {
+
+    }
+
+    @Override
+    public void reconfigure(final Map<String, ?> configs) {
+
+    }
+
+    @Override
+    public void contextChange(final MetricsContext metricsContext) {
+
+    }
+
+    private synchronized void addMetric(
+        final KafkaMetric metric,
+        final String queryId,
+        final String topic
+    ) {
+        final MetricName metricName = getMetricNameWithQueryIdTag(queryId, metric.metricName());
```

Comment:
The `MetricName` for our ksql metrics is mostly the same as the original Streams metric, except that we add the `queryId` in the tags

Reply:
Not sure I understand -- note that we're not technically "changing" or removing the original metrics, we're just adding new metrics that have aggregated over all the taskId & processorNode tags, and this new metric is given a query tag as well

Although that reminds me, the ksql metrics' `MetricName` should not include the `taskId` and `processorNode` tags -- I'll update the code to remove that. Thanks

Reply:
Ah I think I see what you're saying, once I fix the above then yes, there's no reason to add the `queryId` to the `MetricName` of the original metrics from Streams -- only the actual `MetricName` of the aggregated throughput total needs to have this tag. I'll update that as well